### DISCUSSION TITLE
Dynamically Serving Inference Model Adapters with ORT

### DISCUSSION THREAD

**Discussion by contrebande-labs** [microsoft/onnxruntime#21406] [API Q&A]
Hi Guys,

I'm trying to build a inference server in ORT (Java) and want to add dynamic model adapter serving capabilities as in[ this HF blog post here](https://huggingface.co/blog/multi-lora-serving). Specifically I want to be able to dynamically "enable" and "disable" preloaded LoRAs and DoRAs for a UNet. From what I've seen so far, I should be doing this by using fixtures like [`AddExternalInitializers()`](https://onnxruntime.ai/docs/api/c/struct_ort_api.html#a2ac7f2c04d5ee25a05ee8b34307173b0) , [`AddExternalInitializersFromFilesInMemory()`](https://onnxruntime.ai/docs/api/c/struct_ort_api.html#a0ab38aa9431587cb5e0f4301b52e30ae) and [`AddInitializer()`](https://onnxruntime.ai/docs/api/c/struct_ort_api.html#a21a723218a1f795a0deabb6f8a2c541a). The thing is that I have no idea how to use these methods. Can anybody point me to code examples, ideally in Java?

Thanks!

**Comment by Craigacp** [comment]
That's not possible with  ORT as those take effect when the session is initialized. You could potentially add a single empty LoRA to an ONNX UNet model and then add the weights for that as an initializer, but you'd need to destroy the session & recreate it every time you wanted to change LoRA, and you'd also need a different ONNX model for different size LoRAs or ones that are placed on different layers. Creating the ONNX model with the attached LoRA is something I'm not sure how to do automatically, though it's definitely possible if you edit an ONNX protobuf (but it would be nasty to do by hand).

The tests for `AddExternalInitializers` (https://github.com/microsoft/onnxruntime/blob/main/java/src/test/java/ai/onnxruntime/InferenceTest.java#L695) and `AddInitializer` (https://github.com/microsoft/onnxruntime/blob/main/java/src/test/java/ai/onnxruntime/InferenceTest.java#L743) show hot to use it. I don't think I've got around to exposing `AddExternalInitializersFromFilesInMemory` yet as it's not possible to easily mmap large files in Java without Java 22's `MemorySegment`.

**Reply by Craigacp** [reply]
I just want to check that you've seen the stable diffusion implementation that I released as a demo for ORT last year - https://github.com/oracle/sd4j. I think you've probably got all the bits done already, but that's supposed to show how to use a bunch of features. I've not had chance to fully update it for ORT 1.17 onwards yet which will pin all the buffers, but I might find time to do that in the next month.

**Reply by contrebande-labs** [reply]
Yes, that's the code I started from. Love it! Is it possible to initialize a session with a protobuffer graph java object and weights that are already in memory? I actually don't mind operating on the graph in java (and generating code from the onnx proto schemas) but I would like to avoid reloading the weights each time.

EDIT: I'm using Java 22 (and willing to enable all the preview/experimental features if necessary). I'd be willing to work on a public fork of SD4J and do PRs to the original repo if that could help. I've added server that builds a SD4J request from a HTTP request JSON, and generalized the whole thing to serve other models than SD, so I don't know if that fits the original SD4J scope, but let me know if you are interested.

**Reply by Craigacp** [reply]
It is possible to initialize a session from in memory objects, you'll need a byte array for the protobuf and all the initializers are OnnxTensors. If any individual tensor is bigger than 2GB then we've got a problem, but I don't think they should be?

The Java 22 support would need to be in ORT itself, and that we can't move from Java 8 because it also supports Android. Without more resources I can't easily support a Java 22 ORT and a Java 8 one, so things are stuck at the baseline for the moment. When ORT decides to rev the minimum Android version it supports I hope to move it to Java 17, but that's a decision for the ORT PMs.

I think if you want to make a big stable diffusion Java app then that's probably not what SD4J is for because I'd like it to be simple enough to serve as demo code, but adding a few new features is fine. I think the server probably wants to live in another project, but if it needs changes to SD4J to better support the server then maybe some of those should live in SD4J as we do want it to be usable in other projects. Maybe we should follow up on email to discuss?

**Reply by contrebande-labs** [reply]
Yes we can and should talk. Let me just mention here and now that I think the Java ORT bindings should altogether live outside of the ORT main repo if the ORT PMs are not willing to maintain it separately from the Android fork. Android is not Java and the two should never be confused. As I wrote my first lines of code using the Android bindings (because that's how we should refer to them), I immediatly wanted to fork it to add support for the FFM API instead of JNI (and not bundling the compiled libraries in the jars, for instance), the Vector API, Java-native ONNX Protobuf manipulation, etc. The list is long. A while back, I was also interested in IR bytecode (JAX, MLIR, etc.) vs Java bytecode in general and saw a lot of potential for Java in that space. But this has to come from Oracle, not Microsoft. I'm emailing you now.

**Reply by Craigacp** [reply]
It targeted Java 8 originally because I had internal users on 8 who wanted it when I wrote the initial implementation in 2019. I think those internal users have moved to newer versions, but I think that's 17 or 21 not 22, and I don't have bandwidth to maintain multiple versions. MemorySegment is important to support, but the VectorAPI doesn't really add much without surfacing higher level operations, and having a good Tensor API requires some language support which unfortunately isn't present in the Java platform yet.

We have a bunch of Java code for manipulating ONNX protobufs in Tribuo - https://github.com/oracle/tribuo/tree/main/Util/ONNXExport, a longer term goal there is to automatically generate the [ONNX op enum](https://github.com/oracle/tribuo/blob/main/Util/ONNXExport/src/main/java/org/tribuo/util/onnx/ONNXOperators.java) from the ONNX project itself and then we'd have coverage of all the ops rather than just the ones we needed for Tribuo's ONNX export support. It would also be nice to load in an ONNX protobuf and manipulate it using that API which has a lot more safety than working on the protobuf directly, but we've not had time to build that (mostly because we don't have full op support). 

There's also things like [OpenJDK Babylon](https://openjdk.org/projects/babylon/) where it would be fun to compile down to ONNX from a Java method, and I've talked to the Babylon team about doing that at some point. They can target OpenAI's Triton from Bablyon, and I know MLIR has been something they are looking closely at for that project.

**Reply by contrebande-labs** [reply]
I guess, I will try my hands on Java 22 experimental ORT bindings, then. Will also look at Tribuo. Will report progress here. Might take a while.

**Reply by Craigacp** [reply]
There's a direct binding with FFM here - https://github.com/yuzawa-san/onnxruntime-java, but it doesn't conform to the ORT Java API.

**Reply by contrebande-labs** [reply]
I will look at it for sure. But I want it to be a 1:1 equivalent (and drop-in replacement) with both what you've done AND the ORT C/C++ API.
